### ISSUE TITLE
Standardize SDK logging levels

### ISSUE THREAD

**Issue by MrAlias** [open-telemetry/opentelemetry-specification#2039]
Currently, the SDK logging is specified to be configured via the [`OTEL_LOG_LEVEL` environment variable](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.7.0/specification/sdk-environment-variables.md#general-sdk-configuration). There currently is only one log level listed in the specification: `info`. However, other log levels are not specified. It is important that each language use the same logging level so user configuration of environment variables will be cross-language compatible.

The [compliance matrix](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.7.0/spec-compliance-matrix.md#environment-variables) shows only three language implement this environment variable: javascript, ruby, and erlang. Auditing these languages I have found the following logging levels being used.

### Javascript

https://github.com/open-telemetry/opentelemetry-js/blob/feea5167c15c41f0aeedc60959e36c18315c7ede/packages/opentelemetry-core/src/utils/environment.ts#L187-L195

- ALL
- VERBOSE
- DEBUG
- INFO
- WARN
- ERROR
- NONE

### Ruby

https://github.com/open-telemetry/opentelemetry-ruby/blob/62d6ee584ce23ab23d8cde447c872b49486cedbd/sdk/lib/opentelemetry/sdk/forwarding_logger.rb#L20-L31

- debug
- info
- warn
- error
- fatal
- unknown

### Erlang

As far as I can tall (I'm definitely not an Erlang programmer), they are using the built-in logging levels

- emergency 
- alert
- critical
- error
- warning
- notice
- info
- debug

## Proposed Solution

The intersection of the already implemented logging levels are: `debug`, `info`, `error`.

Standardizing on these values would mean that these languages will remain compliant. However, currently, only Javascript has released a stable release of OpenTelemetry. That would mean that if we wanted to add more than this minimal set of logging levels they should ideally remain compatible with the Javascript levels. Specifically, it seems having a "none" level would be ideal for users that do not want any logging output.

**Comment by MrAlias** [comment]
@open-telemetry/javascript-approvers @open-telemetry/ruby-approvers @open-telemetry/erlang-approvers

**Comment by hauleth** [comment]
Erlang uses [RFC 5424](https://datatracker.ietf.org/doc/html/rfc5424) log levels. Elixir before 1.11 used only `debug`, `info`, `warning`, and `error`, but recent versions support all 8 levels.

Technically there are 2 additional configuration options for Erlan `all` and `none`. While `all` is equivalent to `debug` there is difference between `emergency` and `none`, as the first will still pass `emergency` events while `none` will filter out everything (basically it disables logging).

**Comment by pellared** [comment]
> The intersection of the already implemented logging levels are: debug, info, error

I think that `warn` can be safely added to the list.

**Comment by dyladan** [comment]
> While I agree that we should define a common "set" of logging levels that "should" be supported, different runtimes, languages and libraries will have their own set of levels that are used by their communities.
> 
> So as per the ([signalfx/gdi-specification#139](https://github.com/signalfx/gdi-specification/pull/139)), I like the definition of
> 
> `Distribution MAY support additional levels:`

I think I disagree. This is a config SIG issue in my opinion, but what we want is a single configuration that can be deployed to all your SDKs and your collector that behaves in a consistent manner.

**Comment by bryannaegele** [comment]
I think we're good now in Erlang as the levels are getting translated.

**Comment by jack-berg** [comment]
If we add support for log level to file config, we can standardize on an enumeration, and let each language define how values map to their ecosystem. A natural starting point would be to use the severity level definition from the log data model.

**Comment by trentm** [comment]
> If we add support for log level to file config, we can standardize on an enumeration, and let each language define how values map to their ecosystem. A natural starting point would be to use the severity level definition from the log data model.

This came up in a current PR for OTel JS. The PR is attempting to use `log_level` from a declarative config file to set the OTel JS SDK's internal logger:

```js
    diag.setLogger(new DiagConsoleLogger(), { logLevel: config.log_level });
```

For that to work, I think we'd want the declarative config accepted values for `log_level` to include `all` and `none` values (or something equivalent).

Would there be objections to that?

**Comment by jack-berg** [comment]
Right now the declarative config data model uses the same severity number scheme from the log data model:

- Declarative config SeverityNumber type: https://github.com/open-telemetry/opentelemetry-configuration/blob/main/schema-docs.md#severitynumber-
- Log data model severity definition: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#displaying-severity

I think its safe to say that `FATAL4` is equivalent to all. We could say the lowest level, `debug`, is equivalent to none. But that feels odd.

Adding a new `none` to accommodate `.log_level` feels fine enough, but the `SeverityNumber` type is also used in the [`LoggerConfig` type](https://github.com/open-telemetry/opentelemetry-configuration/blob/main/schema-docs.md#experimentalloggerconfig). `none` feels stranger there because there's been all sort of conversations about whether ServerityNumber=0 is allowed.

**Comment by jack-berg** [comment]
BTW, I think we close this issue as "wont fix". Its been open five years, and the chances of us going back and retroactively standardizing `OTEL_LOG_LEVEL` are minimal. We solved the standardization problem in declarative config. The latest comment by @trentm is a separate issue about extending the standard set defined in declarative config.

**Comment by pellared** [comment]
> We solved the standardization problem in declarative config. 

How about **recommending** using the same values as in declarative config? Even if some languages may already use different values they may want to change it in a vNext version.

**Comment by jack-berg** [comment]
Sounds good to me.

**Comment by trentm** [comment]
tl;dr: All good for OTel JS. We can document usage of `log_level: trace` for "ALL" and `log_level: fatal4` for "NONE" and move on.


> I think its safe to say that `FATAL4` is equivalent to all. We could say the lowest level, `debug`, is equivalent to none. But that feels odd.

That is backwards.
- `FATAL4` (or specifically `fatal4` lower-case, as the declarative config schema currently requires it) is equivalent to **none** -- almost (unless some internal SDK logging uses level `fatal4`).
- **`trace`** (not `debug`) is equivalent to `all`.



> Adding a new `none` to accommodate `.log_level` feels fine enough, but the `SeverityNumber` type is also used in the [`LoggerConfig` type](https://github.com/open-telemetry/opentelemetry-configuration/blob/main/schema-docs.md#experimentalloggerconfig). `none` feels stranger there because there's been all sort of conversations about whether ServerityNumber=0 is allowed.

`.log_level` is defining a minimum severity, as is `LoggerConfig.minimum_severity`.
The `SeverityNumber` type *name* in declarative config, is a bit misleading (it isn't a number).
Something like `MininumSeverity` might have been more accurate, but with 1.0 that ship has sailed, IIUC.

The missing `all` value is a user convenience that can be solved with a doc suggestion to use `trace`.

The missing `none` value is a user convenience for *disabling* internal SDK logging. More related to `LoggerConfig.enabled`.
Theoretically one might argue for `logging.enabled` and `logging.minimum_severity` rather than `log_level` in declarative config, but that is probably overkill.
If JS/Node.js wants a way for SDK users to disable internal logging, we can doc using `log_level: fatal4` and move on.
That is weird for a user, but a rare use case.
